xbmc/filesystem/SpecialProtocol.cpp

```diff
@@ -28,12 +28,402 @@
+    else
+      userHome = "/root";
+
+#if defined(TARGET_DARWIN_IOS)
+    path = userHome + "/" + CDarwinUtils::GetAppRootFolder() + "/" + CCompileInfo::GetAppName() + "/userdata";
+#else
+    path = userHome + "/Library/Application Support/" + CCompileInfo::GetAppName() + "/userdata";
+#endif
+  }
+  else
+  {
+    path = URIUtils::AddFileToFolder(GetXBMCPath(), "portable_data");
+  }
+
+#endif
+  return path;
+}
+
+std::string CSpecialProtocol::GetMasterProfilePath(const std::string &fileName)
+{
+  return URIUtils::AddFileToFolder(GetMasterProfilePath(), fileName);
+}
+
+
+std::string CSpecialProtocol::GetXBMCPath()
+{
+  std::string path;
+#if defined (TARGET_WINDOWS)
+  path = CUtil::GetHomePath();
+#endif
+
+#if defined(TARGET_POSIX) && !defined(TARGET_DARWIN)
+  path = INSTALL_PATH;
+  /* Check if binaries and arch independent data files are being kept in
+  * separate locations. */
+  if (!CDirectory::Exists(URIUtils::AddFileToFolder(path, "userdata")))
+  {
+    /* Attempt to locate arch independent data files. */
+    auto appBinPath = CUtil::GetHomePath("KODI_BIN_HOME");
+    path = CUtil::GetHomePath(appBinPath);
+    if (!g_application.PlatformDirectoriesEnabled())
+    {
+      URIUtils::AddSlashAtEnd(path);
+    }
+  }
 #endif
+#if defined (TARGET_DARWIN)
+  path = CUtil::GetHomePath();
+  if (!g_application.PlatformDirectoriesEnabled())
+  {
+    URIUtils::AddSlashAtEnd(path);
+  }
+#endif
+  return path;
+}
+
+std::string CSpecialProtocol::GetXBMCPath(const std::string &fileName)
+{
+  return URIUtils::AddFileToFolder(GetXBMCPath(), fileName);
+}
+
+
+std::string CSpecialProtocol::GetXBMCBinPath()
+{
+  std::string path;
+#ifdef TARGET_WINDOWS
+  path = GetXBMCPath();
+#endif
+
+#if defined(TARGET_POSIX) && !defined(TARGET_DARWIN)
+  path = CUtil::GetHomePath("KODI_BIN_HOME");
+  if (!g_application.PlatformDirectoriesEnabled())
+  {
+    URIUtils::AddSlashAtEnd(path);
+  }
+#endif
+#ifdef TARGET_DARWIN
+  path = GetXBMCPath();
+  if (!g_application.PlatformDirectoriesEnabled())
+  {
+    URIUtils::AddSlashAtEnd(path);
+  }
+#endif
+  return path;
+}
+
+std::string CSpecialProtocol::GetXBMCBinPath(const std::string &fileName)
+{
+  return URIUtils::AddFileToFolder(GetXBMCBinPath(), fileName);
+}
+
+
+std::string CSpecialProtocol::GetXBMCBinAddonPath()
+{
+  std::string path;
+#ifdef TARGET_WINDOWS
+  path = GetXBMCPath("/addons");
+#endif
+#if defined(TARGET_POSIX) && !defined(TARGET_ANDROID)
+  path = GetXBMCBinPath() + "/addons";
+#endif
+#ifdef TARGET_ANDROID
+  path = CXBMCApp::getApplicationInfo().nativeLibraryDir.c_str();
+#endif
+  return path;
+}
+
+std::string CSpecialProtocol::GetXBMCBinAddonPath(const std::string &fileName)
+{
+  return URIUtils::AddFileToFolder(GetXBMCBinAddonPath(), fileName);
+}
+
+
+std::string CSpecialProtocol::GetXBMCAltBinAddonPath()
+{
+  std::string path;
+#if defined(TARGET_POSIX)
+  /* REVIEW:
+  if (getenv("KODI_BINADDON_PATH")) // don't see why to do this check. what to do if false?
```

Comment:
yeah this check seems redundant, and can be removed.

Reply:
I was wrong about this. It returns null if the environment variable doesn't exist and assigning that to a string will cause issues so the check should stay.